Feature/tabular analytics

alitolga opened the pull request:
First implementation for the Tabular Analytics.

Feel free to leave comments/remarks.

alitolga committed:
First commit for Tabular Analytics (mean)

alitolga committed:
Add Unit Tests

srcansiz commented on tests/test_tabular_analytics.py:
```diff
@@ -0,0 +1,234 @@
+    return mock_dataset
+
+
+class TestTabularAnalyticsMean:
```
@alitolga, I think you don't need t use test class here. The functions contains `self` but not used. Please use functions instead of classes

srcansiz commented on fedbiomed/common/analytics/analytics_dispatcher.py:
```diff
@@ -0,0 +1,17 @@
+    @classmethod
+    def mean(self, dataset, params=None):
+        analytics_cls = self.registry[type(dataset)]
+        return analytics_cls().mean(dataset, params)
```
Hi @alitolga,

Thank you, very nice idea to implement `Dispatcher`!

srcansiz replied:
Do you think that this class can be implemented as `Mixin` class. For example, this implementation assumes that the dataset is going to be given as argument. 
```
my_dataset = TabularDataset(...)

# I expect the usage will be
mean = AnalyticsDispatcher.mean(my_dataset, params={}) 

```

But what about making Analytics classes as mixin classes where it can be integrated directly in the dataset

```
class TabularDataset(.., _TabularAnalytics)
....

# Then, the usage will be 

my_dataset = TabularDataset(...)
my_dataset.mean(params)
```

alitolga replied:
I changed the implementation according to the mixin paradigm. I wait for your feedback :)

srcansiz commented on fedbiomed/common/analytics/tabular_analytics.py:
```diff
@@ -0,0 +1,48 @@
+
+
+class TabularAnalytics(AnalyticsStrategy):
+    def mean(self, dataset: TabularDataset, params=None):
```
Can you please add return type.

srcansiz replied:
And `type` for `params` argument

srcansiz commented on fedbiomed/common/analytics/tabular_analytics.py:
```diff
@@ -0,0 +1,48 @@
+        if not isinstance(reader, CsvReader):
+            raise FedbiomedError(
+                f"Expected CsvReader, got {type(reader).__name__} instead."
+            )
```
No need to check reader type it is already checked implicitly by checking the dataset type.

srcansiz commented on fedbiomed/common/analytics/tabular_analytics.py:
```diff
@@ -0,0 +1,48 @@
+        if not isinstance(controller, TabularController):
+            raise FedbiomedError(
+                f"Expected TabularController, got {type(controller).__name__} instead."
+            )
```
Isn't possible to use the `dataset` directly. Here, you are accessing a private member `_controller` which is not an expected way of accessing class member.

srcansiz commented on fedbiomed/common/analytics/tabular_analytics.py:
```diff
@@ -0,0 +1,48 @@
+
+        missing = [c for c in cols if c not in df.columns]
+        if missing:
+            raise KeyError(f"Columns not found: {missing}")
```
Isn't this logic already implemented in `TabularDataset`, like making sure that column required is existing or not , selecting a specific column etc.

srcansiz commented on fedbiomed/common/analytics/tabular_analytics.py:
```diff
@@ -0,0 +1,48 @@
+        if not numeric_cols:
+            return {}
+
+        result = df.select(numeric_cols).mean()
```
I prefer to use `TabularDataset` directly rather than accessing `df` from controller.

alitolga committed:
Change the implementation according to mixin paradigm.

- Remove analytics dispatcher.
- Move mean calculation to base analytics strategy class.
- Add Tabular Analytics mixin class to Tabular Dataset.
- Write unit tests.

srcansiz commented on fedbiomed/common/analytics/analytics_strategy.py:
```diff
@@ -0,0 +1,32 @@
+        data_mean = data_sum / count
+        target_mean = target_sum / count
+
+        return {"data": data_mean, "target": target_mean}
```
Hi @alitolga,

I think for the first version it could be nice to have only `TabularAnalytics` without base class. It is because we don't know which parts can be shared among datasets and analytics. Therefore, it is logic to have simple implementation first and improve it as we add other analytics.

alitolga replied:
Hi @srcansiz ,

I applied the suggested changes.

srcansiz commented on fedbiomed/common/analytics/analytics_strategy.py:
```diff
@@ -0,0 +1,32 @@
+        data_mean = data_sum / count
+        target_mean = target_sum / count
+
+        return {"data": data_mean, "target": target_mean}
```
Another thing, I think that `target` does not make sense for analytics. Can you please check if it is possible to create a dataset `TabularDataset` without specifying `target` argument, if not propose a version where it is possible. This will allow to use only `data` for analytics.

alitolga replied:
Hi @srcansiz ,

The analytics return a dictionary with the column names now.

alitolga committed:
Apply suggested changes

srcansiz commented on fedbiomed/common/analytics/tabular_analytics.py:
```diff
@@ -0,0 +1,60 @@
+        for i, col in enumerate(self._target_columns):
+            result[col] = (
+                target_mean[i] if hasattr(target_mean, "__getitem__") else target_mean
+            )
```
Hi @alitolga, 

How do you handle the case where there is no target columns provided? Given a target for `TabularDataset` in analytics does not make sense no? It may require to to modify `TabularDataset` implementations to provide columns only without separating `inputs` and `targets`.

srcansiz replied:
or using only `inputs` and not forcing `target` to be provided.

alitolga replied:
Hi @srcansiz ,

I have done the necessary modifications.

alitolga committed:
Change target columns parameter to Optional for Tabular Dataset. Modify necessary functions

alitolga committed:
Add support for tabular analytics when target is None

srcansiz commented on fedbiomed/common/analytics/tabular_analytics.py:
```diff
@@ -0,0 +1,65 @@
+        """
+        data_sum = 0
+        target_sum = 0
+        count = 0
```
Hi @alitolga, 

The implementation is lacking validation of column types. Single this classes intended to be used with `TabularDataset`, I think you can acces `self._controller` to validate if given columns are numeric not string.

alitolga replied:
Hi @srcansiz ,

I added the function to filter the non-numeric values using the schema.

srcansiz commented on fedbiomed/common/analytics/tabular_analytics.py:
```diff
@@ -0,0 +1,65 @@
+            if self._target_columns is not None:
+                for col in self._target_columns:
+                    result[col] = None
+            return result
```
I think you can return directly raise an error if there is no sample. You can do it in the very beginning of the method.
```
if not len(self):
    raise FedbiomedError("....")
```

srcansiz commented on fedbiomed/common/analytics/tabular_analytics.py:
```diff
@@ -0,0 +1,65 @@
+        for idx in range(len(self)):
+            data, target = self[idx]
+            data_sum += data
+            count += 1
```
No need to have `count`. It is equal to `len(self)`

srcansiz commented on fedbiomed/common/analytics/_tabular_analytics.py:
```diff
@@ -0,0 +1,65 @@
+        # Add input column means
+        for i, col in enumerate(self._input_columns):
+            result[col] = (
+                data_mean[i] if hasattr(data_mean, "__getitem__") else data_mean
```
Is `hasattr` check really needed here ?

alitolga replied:
It is just to be error-proof in case there is only one input column and data_mean returns a scalar.

alitolga committed:
Add check to filter only numeric columns

alitolga committed:
Remove counter

srcansiz committed:
renames analytics modules and implements minor changes
